internal/typeinfo/typeinfo_test.go

```diff
@@ -0,0 +1,227 @@
+		index:     3,
+		omitEmpty: false,
+		tag:       "IdENT99",
+	}, {
+		fieldName: "ValidTag1",
+		index:     2,
+		omitEmpty: false,
+		tag:       "_i_d_55_",
+	}, {
+		fieldName: "ID",
+		index:     0,
+		omitEmpty: false,
+		tag:       "id",
+	}, {
+		fieldName: "Name",
+		index:     1,
+		omitEmpty: true,
+		tag:       "name",
+	}}
+
+	allOutputs, names, err := argInfo.AllStructOutputs("myStruct")
+	c.Assert(err, IsNil)
+	c.Assert(allOutputs, HasLen, len(structFields))
+
+	for i, t := range structFields {
+		expectedStructField := &structField{
+			name:       t.fieldName,
+			structType: reflect.TypeOf(myStruct{}),
+			index:      t.index,
+			tag:        t.tag,
+			omitEmpty:  t.omitEmpty,
+		}
+
+		input, err := argInfo.InputMember("myStruct", t.tag)
+		c.Assert(err, IsNil)
+		c.Assert(input, DeepEquals, expectedStructField)
+
+		output, err := argInfo.OutputMember("myStruct", t.tag)
+		c.Assert(err, IsNil)
+		c.Assert(output, DeepEquals, expectedStructField)
+
+		c.Assert(allOutputs[i], DeepEquals, expectedStructField)
+		c.Assert(names[i], Equals, t.tag)
+	}
+}
+
+func (s *typeInfoSuite) TestArgInfoMap(c *C) {
+	type myMap map[string]any
+
+	argInfo, err := GenerateArgInfo([]any{myMap{}})
+	c.Assert(err, IsNil)
+
+	expectedMapKey := &mapKey{mapType: reflect.TypeOf(myMap{}), name: "key"}
+
+	input, err := argInfo.InputMember("myMap", expectedMapKey.name)
+	c.Assert(err, IsNil)
+	c.Assert(input, DeepEquals, expectedMapKey)
+
+	output, err := argInfo.OutputMember("myMap", expectedMapKey.name)
+	c.Assert(err, IsNil)
+	c.Assert(output, DeepEquals, expectedMapKey)
+}
+
+// This struct is used to test shadowed types in TestGenerateArgInfoInvalidTypeErrors
+type T struct{ foo int }
+
+var t = T{}
+
+func (s *typeInfoSuite) TestGenerateArgInfoInvalidTypeErrors(c *C) {
+	type T struct{ foo int }
+	type M map[string]any
+
+	_, err := GenerateArgInfo([]any{nil})
+	c.Assert(err, ErrorMatches, "need struct or map, got nil")
+
+	_, err = GenerateArgInfo([]any{struct{ foo int }{}})
+	c.Assert(err, ErrorMatches, "cannot use anonymous struct")
+
+	_, err = GenerateArgInfo([]any{map[string]any{}})
+	c.Assert(err, ErrorMatches, "cannot use anonymous map")
+
+	_, err = GenerateArgInfo([]any{T{}, T{}})
+	c.Assert(err, ErrorMatches, `found multiple instances of type "T"`)
+
+	_, err = GenerateArgInfo([]any{(*T)(nil)})
+	c.Assert(err, ErrorMatches, "need struct or map, got pointer to struct")
+
+	_, err = GenerateArgInfo([]any{(*M)(nil)})
+	c.Assert(err, ErrorMatches, "need struct or map, got pointer to map")
+
+	_, err = GenerateArgInfo([]any{""})
+	c.Assert(err, ErrorMatches, "need struct or map, got string")
+
+	_, err = GenerateArgInfo([]any{0})
+	c.Assert(err, ErrorMatches, "need struct or map, got int")
+
+	_, err = GenerateArgInfo([]any{[10]int{}})
+	c.Assert(err, ErrorMatches, "need struct or map, got array")
+
+	_, err = GenerateArgInfo([]any{t, T{}})
+	c.Assert(err, ErrorMatches, `two types found with name "T": "typeinfo.T" and "typeinfo.T"`)
```

Comment:
Can we refactor this to be a table and run the test against each test case in the table?